alacritty_terminal/src/tty/windows/conpty.rs

```diff
@@ -230,6 +241,58 @@ pub fn new(config: &Options, window_size: WindowSize) -> Option<Pty> {
     Some(Pty::new(conpty, conout, conin, child_watcher))
 }
 
+// Windows environment variables are case-insensitive, and the caller is responsible for
+// deduplicating environment variables, so do that here while converting.
+//
+// https://learn.microsoft.com/en-us/previous-versions/troubleshoot/windows/win32/createprocess-cannot-eliminate-duplicate-variables#environment-variables
+fn convert_custom_env(custom_env: &HashMap<String, String>) -> Option<Vec<u16>> {
+    // Windows inherits parent's env when no `lpEnvironment` parameter is specified.
+    if custom_env.is_empty() {
+        return None;
+    }
+
+    let mut converted_block = Vec::new();
+    let mut all_env_keys = HashSet::new();
+    for (custom_key, custom_value) in custom_env {
+        let custom_key = OsStr::new(custom_key);
+        if all_env_keys.insert(custom_key.to_ascii_uppercase()) {
```

Comment:
`custom_env` is a `HashMap`, so `custom_key` is known to not exist in `all_env_keys`. What is the point of checking if it was already inserted? And even if it was, wouldn't we want to overwrite it rather than use the *first* value specified?

Reply:
> What is the point of checking if it was already inserted

To deduplicate it case-insensitively: `HashMap::from([("one_key", "one_value"), ("ONE_KEY", "another value")])` will contain two keys but from Windows' standpoint, both keys are the same.

> wouldn't we want to overwrite it rather than use the first value specified

Since it's a `HashMap`, is there such thing as the first value, really? 

I have thought that it's dangerous to rely on hash-based collections' order of iteration, so we cannot safely state that the user env input deserialized into a `HashMap` will have the same order of entries.

What we do want, is to accept user only env vars in case of key matching case-insensitively, so I have used the same code for both to make it look less bloated.

------------------

I'm somewhat disenchanted now: there's a comment above, saying 
```
// Windows environment variables are case-insensitive, and the caller is responsible for
// deduplicating environment variables, so do that here while converting.
```
and here's the `to_ascii_uppercase` call — should I remove the entire doc then, if it's useless? Or adjust somehow?

Reply:
> and here's the to_ascii_uppercase call — should I remove the entire doc then, if it's useless? Or adjust somehow?

Nah it's fine, I just didn't consider the implications of this for the initial hashmap.

It might make sense to warn in the else case though, which also would make it more clear to anyone reading the code? People should be made aware that environment variables they're defining have no effect, though maybe that should just be a debug statement for configs shared with linux?

Reply:
Yeah, somewhat tricky to log it for all cases, I pushed a change to WARN log only windows part only for duplicates in the user config.

* Unix part with case sensitivity makes `HashMap<String, String>` interface "intuitive" and all deduplications already happen in the user code
* Both Windows and Unix versions are expected to deduplicate OS env vars, so no logging is needed there
* The "unintuitive" case-insensitive Windows part got logs for the rest